### PR TITLE
Fix order bug in optimizer.

### LINKED ISSUE #7411: [yul] optimizer can produce different side-effects because of the order of evaluation of for-loop conditional expression that contains `msize()` or a user-defined function that has a side-effect
## Description

```
{
  for {} div(create(0, 1, 0), shl(msize(), 1)) {}
  {
  }
}
```

Not sure why the yul interpreter is evaluating create twice (trace below)

```
Trace:
  CREATE(0, 1, 0)
  CREATE(0, 1, 0)
```

while the optimized version of the program.

```
{
    for { } shr(msize(), create(0, 1, 0)) { }
    { }
}
```

reports this trace

```
Trace:
  CREATE(0, 1, 0)
```

tripping the fuzzer in the process.

The rule list responsible for this is the following
https://github.com/ethereum/solidity/blob/324cc71b13a2e1ebf1fe34397d085832b89076a4/libevmasm/RuleList.h#L489-L496
## Environment

- Compiler version: latest develop

## Steps to Reproduce

```
$ yulrun --input-file <test.yul>
...
$ solc --strict-assembly --optimize <test.yul>
$ yulrun --input-file <test_optimized.yul>
```

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/7411

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -13,6 +13,7 @@ Compiler Features:
 
 
 Bugfixes:
+ * Yul Optimizer: Fix reordering bug in connection with shift instructions in for loop conditions.
```
Comment (bshastry) [inline Changelog.md]: Edit: In connection with division or multiplication combined with right/left shifted one in for loop conditions?

**Review by bshastry — DISMISSED** [review]
LGTM apart from a minor comment on change log description. Commits can be squashed I feel.

File: Changelog.md (hunk 1 above)
Comment (bshastry) [reply Changelog.md]: Also, did we decide to skip a bug list entry for this?

**Review by bshastry — APPROVED** [review]
LGTM. I don't have a strong opinion about this making to the bug list since it appears rather unlikely that the buggy pattern is hand-written. I am not very sure of this so it is better another person confirms this.

**Comment by chriseth** [comment]
Actually I was not even aware in the beginning about how specific this pattern is.

Note that you need ``div(x, shl(y, 1))`` where neither `x` nor `y` can be constants or variables and both have to be functions or opcodes with side-effects that conflict when swapped and this expression needs to be part of a loop condition...

**Pull request merged**